### PR TITLE
Http Proxy raw channel establishment API

### PR DESCRIPTION
**Pull request by bretambrose**
* Reworks http proxy layer to use pass-through handlers rather than stripping or re-using the proxy handler
* Adds an API to create a raw channel through an http proxy.  

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: include/aws/http/private/h1_connection.h [hunk 1]
```diff
@@ -198,4 +198,17 @@ void aws_h1_connection_try_write_outgoing_stream(struct aws_h1_connection *conne
  */
 void aws_h1_connection_try_process_read_messages(struct aws_h1_connection *connection);
 
+/**
+ * Validate and perform a protocol switch on a connection.  Protocol switching essentially turns the connection's
+ * handler into a dummy pass-through.  It is valid to switch protocols to the same protocol resulting in a channel
+ * that has a "dead" http handler in the middle of the channel (which negotiated the CONNECT through the proxy) and
+ * a "live" handler on the end which takes the actual http requests.  By doing this, we get the exact same
+ * behavior whether we're transitioning to http or any other protocol: once the CONNECT succeeds
```
Comment (justinboswell) [inline include/aws/http/private/h1_connection.h:206]: This is a great explanation, thank you!

File: include/aws/http/proxy.h [hunk 2]
```diff
@@ -488,6 +488,19 @@ void aws_http_proxy_options_init_from_config(
     struct aws_http_proxy_options *options,
     const struct aws_http_proxy_config *config);
 
+/**
+ * Establish an arbitrary protocol connection through an http proxy via tunneling CONNECT.  Alpn is
+ * not required for this connection process to succeed, but we encourage it's use if available.
```
Comment (justinboswell) [inline include/aws/http/proxy.h:493]: its

**Review by justinboswell — APPROVED** [review]
LGTM, should really get @graebm or @TingDaoK to take a look though

**Review by graebm — APPROVED** [review]
fix & ship!
nice & clean!

**Pull request merged**